dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,418 @@
+     *
+     * @param bb the block
+     * @return the last node of this block or {@code null}
+     */
+    protected Node getLastNode(Block bb) {
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                return blockContents.get(blockContents.size() - 1);
+            case EXCEPTION_BLOCK:
+                return ((ExceptionBlock) bb).getNode();
+            default:
+                return null;
+        }
+    }
+
+    /**
+     * Generate the order of processing blocks.
+     *
+     * @param cfg the current control flow graph
+     * @return an IdentityHashMap that maps from blocks to their orders
+     */
+    protected IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
+        IdentityHashMap<Block, List<Integer>> depthFirstOrder = new IdentityHashMap<>();
+        int count = 1;
+        for (Block b : cfg.getDepthFirstOrderedBlocks()) {
+            depthFirstOrder.computeIfAbsent(b, k -> new ArrayList<>());
+            depthFirstOrder.get(b).add(count++);
+        }
+        return depthFirstOrder;
+    }
+
+    @Override
+    public String visualizeStore(S store) {
+        return store.visualize(this);
+    }
+
+    /**
+     * Generate the String representation of the nodes of a control flow graph.
+     *
+     * @param visited the set of the visited blocks
+     * @param cfg the control flow graph
+     * @param analysis the current analysis
+     * @return the String representation of the nodes
+     */
+    protected abstract String generateNodes(
+            Set<Block> visited, ControlFlowGraph cfg, @Nullable Analysis<A, S, T> analysis);
+
+    /**
+     * Generate the String representation of an edge.
+     *
+     * @param sId the ID of current block
+     * @param eId the ID of successor block
+     * @param flowRule the content of the edge
+     * @return the String representation of the edge
+     */
+    protected abstract String addEdge(long sId, long eId, String flowRule);
+
+    /**
+     * Return the header of the generated graph. Called by {@link
+     * #visualizeGraphWithoutHeaderAndFooter(ControlFlowGraph, Block, Analysis)}.
+     *
+     * @return the String representation of the header of the control flow graph
+     */
+    protected abstract String visualizeGraphHeader();
+
+    /**
+     * Return the footer of the generated graph. Called by {@link
+     * #visualizeGraphWithoutHeaderAndFooter(ControlFlowGraph, Block, Analysis)}.
+     *
+     * @return the String representation of the footer of the control flow graph
+     */
+    protected abstract String visualizeGraphFooter();
+
+    /**
+     * Return the simple String of the process order of a node. Called by {@link #generateNodes(Set,
+     * ControlFlowGraph, Analysis)}.
+     *
+     * @param order the list of the process order to be processed.
+     * @return the String representation of the process order of the node
```

Comment:
Document the format of the output.  Since the input is a list, could this be a sequence of numbers?  In that case, what order are nodes processed in?

Reply:
Yes, it actually is a sequence of numbers by this version of code.

For example:
`Process order: 23,25`

I think the block with multiple process orders will not be processed multiple times, it is only because that `getDepthFirstOrderedBlocks()` returns it twice. Also, it will be visualized only one time. But, we have two numbers when printing process orders.

Reply:
Are the multiple numbers in the process order ever used?  From what you said, it sounds like they are not.

Reply:
If we pass the argument "verbose" to the program, then the program will print the process order.

I tested it with a program, and some blocks actually have two process orders. Although I don't know if two numbers will be useful, but it indeed happens.

Here is an example of output:
```
285:
Process order: 23, 25
Before: then=org.checkerframework.checker.nullness.NullnessStore (
  all > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=boolean}
  i > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=java.lang.String}
  c > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=char}
  val.length() > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val.charAt(i) > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=char}
  initialized fields = []
  invariant fields = {}
  isPolyNonNull = false
), else=org.checkerframework.checker.nullness.NullnessStore (
  all > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=boolean}
  i > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=java.lang.String}
  c > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=char}
  val.length() > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val.charAt(i) > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=char}
  initialized fields = []
  invariant fields = {}
  isPolyNonNull = false
)
~~~~~~~~~
((WideningConversion(c, int) == WideningConversion(0, int)) || (WideningConversion(c, int) == WideningConversion(1, int)))   [ ConditionalOr ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=boolean}
all = ((WideningConversion(c, int) == WideningConversion(0, int)) || (WideningConversion(c, int) == WideningConversion(1, int)))   [ Assignment ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=boolean}
i   [ LocalVariable ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
1   [ IntegerLiteral ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
(i + 1)   [ NumericalAddition ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
i = (i + 1)   [ Assignment ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
tempPostfix#num1   [ VariableDeclaration ]
tempPostfix#num1   [ LocalVariable ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.UnknownInitialization, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
tempPostfix#num1 = i   [ Assignment ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
tempPostfix#num1   [ LocalVariable ] > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
~~~~~~~~~
After: org.checkerframework.checker.nullness.NullnessStore (
  all > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=boolean}
  i > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=java.lang.String}
  c > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=char}
  tempPostfix#num1 > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  val.length() > CFAbstractValue{annotations=[@org.checkerframework.checker.initialization.qual.Initialized, @org.checkerframework.checker.nullness.qual.NonNull], underlyingType=int}
  initialized fields = []
  invariant fields = {}
  isPolyNonNull = false
)
```

Reply:
Thanks!  Could you add a brief note indicating what you learned?
Then that is more than enough for this pull request.
Thanks a lot for your investigation!